node/pkg/tss/implementation.go

```diff
@@ -0,0 +1,381 @@
+	return &tss.PartyID{
+		MessageWrapper_PartyID: &tss.MessageWrapper_PartyID{
+			Id:      pid.Id,
+			Moniker: pid.Moniker,
+			Key:     pid.Key,
+		},
+		Index: int(pid.Index),
+	}
+}
+func partyIdToProto(pid *tss.PartyID) *gossipv1.PartyId {
+	return &gossipv1.PartyId{
+		Id:      pid.Id,
+		Moniker: pid.Moniker,
+		Key:     pid.Key,
+		Index:   uint32(pid.Index),
+	}
+}
+func (t *Engine) intoGossipMessage(m tss.Message) (*gossipv1.GossipMessage, error) {
+	bts, routing, err := m.WireBytes()
+	if err != nil {
+		return nil, err
+	}
+
+	indices := make([]*gossipv1.PartyId, 0, len(routing.To))
+	for _, pId := range routing.To {
+		indices = append(indices, partyIdToProto(pId))
+	}
+
+	msgToSend := &gossipv1.SignedMessage{
+		Payload:         bts,
+		Sender:          partyIdToProto(m.GetFrom()),
+		Recipients:      indices,
+		MsgSerialNumber: atomic.AddUint64(&t.msgSerialNumber, 1),
+		Authentication:  nil,
+	}
+
+	tssMsg := &gossipv1.PropagatedMessage{}
+
+	if routing.IsBroadcast || len(routing.To) == 0 || len(routing.To) > 1 {
+		t.sign(msgToSend)
+		tssMsg.Payload = &gossipv1.PropagatedMessage_Echo{
+			Echo: &gossipv1.Echo{
+				Message:   msgToSend,
+				Signature: nil, //  No sig here, it means this is the original sender of the message, and not a vote.
+				Echoer:    0,   // TODO: use -1 since this is not an echo.
+			},
+		}
+	} else {
+		t.encryptAndMac(msgToSend)
+		// encrypt then mac the msgToSend (payload encrypted, and mac the full &gossipv1.SignedMessage)
+		tssMsg.Payload = &gossipv1.PropagatedMessage_Unicast{
+			Unicast: msgToSend,
+		}
+	}
+
+	return &gossipv1.GossipMessage{
+		Message: &gossipv1.GossipMessage_TssMessage{
+			TssMessage: tssMsg,
+		},
+	}, nil
+}
+
+func (t *Engine) HandleIncomingTssMessage(msg *gossipv1.GossipMessage_TssMessage) {
+	if t == nil {
+		return
+	}
+
+	defer func() {
+		// todo: consider sending the message to be gossiped or not. perhaps it's a duplicate message?
+	}()
+
+	switch m := msg.TssMessage.Payload.(type) {
+	case *gossipv1.PropagatedMessage_Unicast:
+		parsed, err := t.handleUnicast(m)
+		if err != nil {
+			return
+		}
+
+		// TODO: add the uuid of this message to the set of received messages.
+		t.fp.Update(parsed)
+	case *gossipv1.PropagatedMessage_Echo:
+		parsed, err := t.handleEcho(m)
+		if err != nil {
+			return
+		}
+
+		// fmt.Printf("guardian %v received from %v: %v\n", t.GuardianStorage.Self.Index, parsed.GetFrom().Index, parsed.Type())
+		t.fp.Update(parsed)
+	}
+}
+
+func (t *Engine) handleEcho(m *gossipv1.PropagatedMessage_Echo) (tss.ParsedMessage, error) {
+	echoMsg := m.Echo
+	if echoMsg == nil {
+		return nil, fmt.Errorf("echo message is nil")
+	}
+
+	senderPid := protoToPartyId(echoMsg.Message.Sender)
+	if !t.GuardianStorage.contains(senderPid) {
+		return nil, fmt.Errorf("sender is not a known guardian")
+	}
+
+	if err := t.verifySignedMessage(echoMsg.Message); err != nil {
+		return nil, err
+	}
+
+	if echoMsg.Echoer > uint32(len(t.Guardians)) {
```

Comment:
sholdn't this be >=?

Reply:
it should, but i'm not certain how valuable the change is, since it'll soon change drastically. (relbroadcast branch).